### ISSUE TITLE
Official support for Python 3.12

### ISSUE THREAD

**Issue by SnoopJ** [sopel-irc/sopel#2457]
### Requested Feature

Python 3.12 has [entered beta](https://www.python.org/downloads/release/python-3120b1/), so it's a good time to start thinking about Sopel on that version.

* [ ] Check Sopel tests against 3.12
  * Installing `dev-requirements.txt` is currently blocked by [lack of support](https://github.com/python-greenlet/greenlet/issues/357) in `greenlet`, but there's a PR open for it. I'm able to install from that PR and proceed with the rest of testing
  * Looks like we may need to adjust the version requirement for `setuptools`, I see the following error with `setuptools == 67.8.0` but pinning `setuptools < 67` in `build-system.requires` resolves it.  
  ```
  AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?
  ```
  * There is additional upstream breakage [in `aiohttp`](https://github.com/aio-libs/aiohttp/issues/7229) and [in `frozenlist`](https://github.com/aio-libs/frozenlist/issues/433), seems that the latter has been fixed on development, but there is no PR for `aiohttp` yet. These are both induced by the dependency on `geoip`.
    * This is the point where I gave up trying to run the tests, for the time being :stuck_out_tongue: 
* [ ] Update CI to run with 3.12
  * Looks like the `python-versions` Action is [capable of providing the 3.12.0b1 pre-release](https://github.com/actions/python-versions/blob/7693cef3461cf6902845be55593a2328fef3bea3/versions-manifest.json#L3), so we could run against the beta version(s) before the official release ([scheduled](https://peps.python.org/pep-0693/) for 2 October 2023).
* [ ] Update classifiers list in `pyproject.toml`

**Comment by Exirel** [comment]
Just a note here: there are feature we use in Py3.7 that are deprecated in Py3.9, Py3.10, or Py3.11, and may be removed in version as early as Py3.10 and Py3.11. So I'm pretty sure it'll be quite hard to support such a big range of Python version.

I believe it will be more reasonable to drop anything below Py3.9 to support Py3.12.

**Comment by SnoopJ** [comment]
3.7 reaches its EOL in the next month, so that one is basically free. If we can continue to support 3.8, that would be nice to our downstream, although I don't know what similar issues there might be for that version.

**Comment by SnoopJ** [comment]
This PR is probably a better thing to track for the upstream blockage from `aiohttp` than the issue linked above: https://github.com/bartfeenstra/betty/pull/1008

This issue is still blocked, but it [sounds like](https://github.com/aio-libs/aiohttp/issues/7229#issuecomment-1650308415) 3.12 support might be in the next release.

**Comment by SnoopJ** [comment]
The entire Sopel test suite passes on 3.12.0rc1 with a bit of hand-holding.

```
1317 passed, 8 xfailed, 1 warning in 39.80s
```

Recipe to get a Sopel install on this version:

* Install `greenlet` from source (`pip install git+https://github.com/python-greenlet/greenlet`) ([c2416a1](https://github.com/python-greenlet/greenlet/commit/c2416a16607c4662b41a26a6ad7a2c3a58234ac5))
* Install `aiohttp` from source (clone and run `make`, see instructions here: https://docs.aiohttp.org/en/stable/contributing.html) ([cf97e5b](https://github.com/aio-libs/aiohttp/commit/cf97e5b0a20de31aeb51684add61b7dad4e92375))
  * NOTE: I had some checksum integrity problems with this project's dependencies, but manually dropping the problematic checks in the `npm` machinery got me a working build. Not an issue for us, building a real release wheel is _their_ problem.
* Install `geoip2` from source (clone, edit `setup.cfg` to relax `aiohttp` requirement, and install) ([7769666](https://github.com/maxmind/GeoIP2-python/commit/776966657cdae08aa6c6221f568e9a223acccab8))
* Install Sopel from source (with the `build-system.requires` updated to `setuptools>=63.0,<67`) ([bc688b4](https://github.com/sopel-irc/sopel/commit/bc688b4eaee25a1be4fef66477f016bc21ea61d8))

So things are a looking good for adding Python 3.12 support, we're just waiting for new upstream releases from `greenlet, aiohttp, geoip2` before we're ready to do the rest of the chores.

**Comment by Exirel** [comment]
@SnoopJ this is nice! Thank you very much for that! ❤️

**Comment by dgw** [comment]
"The time has come," the walrus said, "to test on three-point-twelve!"

<img width="326" alt="image" src="https://github.com/sopel-irc/sopel/assets/164140/88128e1a-1406-4450-952d-969a538bb6e7">

_Translation: Python 3.12 went stable today. Blame @xnaas for reminding me, else this ticket would've stayed silent._

However, the [actions/python-versions run](https://github.com/actions/python-versions/actions/runs/6381888192) that was supposed to generate 3.12 failed with errors, so the stable release won't be available through GHA for some unspecified amount of time until the build is fixed.

We probably need to wait on upstream dependencies to fix some number of incompatibilities with the new version anyway before we can officially declare support, anyway. Once GHA gets 3.12 up, we can look into the packages @SnoopJ investigated that were awaiting a 3.12-compatible release as of late August.

**Comment by SnoopJ** [comment]
> However, the [actions/python-versions run](https://github.com/actions/python-versions/actions/runs/6381888192) that was supposed to generate 3.12 failed with errors, so the stable release won't be available through GHA for some unspecified amount of time until the build is fixed.

If I understand right, the thing that failed was the MacOS build specifically? I'm not sure if this failure means that the action we need from GHA isn't available at all or not, that's beyond my ken.

---

I installed 3.12.0 and to my surprise, we seem to have no upstream blockage. I have no idea what changed with `aiohttp` because they definitely haven't done a release with the fix we previously needed, but the only thing I had to do to get the test suite (and my reference test instance) running was to adjust the `setuptools` build requirement. I guess maybe the source of that dependency shifted in the interim.

I'll start a branch later this week for that change, the metadata churn, and whatever needs to be done for CI, but things look pretty promising from a quick check.

**Comment by dgw** [comment]
> If I understand right, the thing that failed was the MacOS build specifically? I'm not sure if this failure means that the action we need from GHA isn't available at all or not, that's beyond my ken.

The workflow graph (see below) indicates that any failing build skips testing/publishing any of the successful builds too, so 3.12 stable won't be available yet in GHA and there hasn't been another build attempt yet that I can see.

<img width="933" alt="image" src="https://github.com/sopel-irc/sopel/assets/164140/17e5d3de-f00c-4899-a957-b23a217cdc00">

**Comment by dgw** [comment]
> [ ] Check on 3.12 support concerns for official non-core plugins

Is this realistically something that even belongs in the core issue tracker?

Presumably only plugins that have dependencies beyond Sopel itself _might_ be affected. The only one we know of for sure is `sopel-iplookup`, which has an install workaround with `--pre` until `aiohttp` gets together a fully 3.12-compatible release.

Everything else, if it doesn't have an 8.0-tested release yet (i.e. still requires `sopel>=xx,<8`), we can test on 3.12 as part of that update. That includes `github` (depends on `bottle`), and `youtube` (depends on `google-api-python-client`) still needs converting to an entrypoint package. The few "official" (or official-ish) plugins that have already been formally bumped to 8.0-compatibility (like `twitter`) should get sorted out through user bug reports as py3.12 adoption grows.

I guess when one of us is bored, we can pick a plugin from the org's repo list and run down a checklist ensuring things like: the plugin 1) is in entrypoint style, 2) works against latest 8.0.0.dev0, and 3) is installable/functional with both py3.8 and py3.12. A ton of the older ones still need converting to entrypoint with `sopel-<pluginname>` package names, and a shim dropped at the old namespace-package location on PyPI; just testing 3.12 then is the most logical approach.

**Comment by Exirel** [comment]
Nah, I don't think we should track that in core, for most plugin except the sopel-help plugin, which is kind of a big deal.

However, I'm pretty sure sopel-help shouldn't have a problem, as long as requests works fine.
